### PR TITLE
enable locking for artifactory backend

### PR DESCRIPTION
**Pull request by bzcnsh**
add support for optional locking with artifactory.
add instructions for artifactory and nexus.

### CODE REVIEW

**Comment by bzcnsh** [comment]
@pselle @jbardin saw your names frequently on recent commits and assumed you have the rights to review PRs. I know you guys are probably extremely busy but can one of you give a quick review to this PR? and let me know if I do anything to expedite its approval.

Thank you!

File: backend/remote-state/artifactory/backend.go [hunk 1]
```diff
@@ -17,29 +18,74 @@ func New() backend.Backend {
...
+				DefaultFunc: schema.EnvDefaultFunc("LOCK_ARTIFACTORY_USERNAME", nil),
+				Description: "username for lock file creation",
+			},
+			"lock_password": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				DefaultFunc: schema.EnvDefaultFunc("LOCK_ARTIFACTORY_PASSWORD", nil),
+				Description: "password for lock file creation",
+			},
+			"unlock_username": &schema.Schema{
```
Comment (jbardin) [inline backend/remote-state/artifactory/backend.go]: We would prefer that the locking mechanism use the same backend credentials as the primary storage. If there is a good reason to separate them, we would want to ensure that the same credentials can both lock and unlock.

File: backend/remote-state/artifactory/backend.go [hunk 2]
```diff
@@ -17,29 +18,74 @@ func New() backend.Backend {
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The repository for lock file access",
+			},
+			"lock_subpath": &schema.Schema{
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "path within the repository for lock file access",
+			},
+			"lock_readback_wait": &schema.Schema{
```
Comment (jbardin) [inline backend/remote-state/artifactory/backend.go]: We prefer not to present implementation details to the use as configuration options.

File: backend/remote-state/artifactory/client.go [hunk 3]
```diff
@@ -2,25 +2,59 @@ package artifactory
...
 
+func logStart(action string) {
+	log.Printf("[TRACE] backend/remote-state/artifactory: starting %s operation", action)
+}
+
+func logResult(action string, err *error) {
+	if *err == nil {
+		log.Printf("[TRACE] backend/remote-state/artifactory: exiting %s operation with success", action)
+	} else {
+		log.Printf("[TRACE] backend/remote-state/artifactory: exiting %s operation with failure", action)
```
Comment (jbardin) [inline backend/remote-state/artifactory/client.go]: the error is not getting logged

File: backend/remote-state/artifactory/client.go [hunk 4]
```diff
@@ -48,16 +79,88 @@ func (c *ArtifactoryClient) Get() (*remote.Payload, error) {
...
+func (c *ArtifactoryClient) Lock(info *state.LockInfo) (string, error) {
+	if c.lockUrl == "" {
+		return "", nil
+	}
+	c.lockID = ""
+	var err error
+	var output []byte
+	logStart("Lock")
+	defer logResult("Lock", &err)
+	rand.Seed(time.Now().UnixNano())
```
Comment (jbardin) [inline backend/remote-state/artifactory/client.go]: A PRNG should only be seeded once, this makes the output more predictable. The provided `info` should already contain a cryptographically secure ID.

File: backend/remote-state/artifactory/client.go [hunk 5]
```diff
@@ -48,16 +79,88 @@ func (c *ArtifactoryClient) Get() (*remote.Payload, error) {
...
+	p := fmt.Sprintf("%s/%s/%s", c.lockRepo, c.lockSubpath, ARTIF_TFLOCK_NAME)
+	if _, err = c.lockNativeClient.Put(p, string(jsonLockInfo), make(map[string]string)); err == nil {
+		if c.lockReadbackWait <= 0 {
+			c.lockID = info.ID
+			c.jsonLockInfo = jsonLockInfo
+			return info.ID, nil
+		}
+		time.Sleep(time.Duration(c.lockReadbackWait) * time.Millisecond)
+		// readback and compare with original info
+		output, err = c.lockNativeClient.Get(p, make(map[string]string))
```
Comment (jbardin) [inline backend/remote-state/artifactory/client.go]: What is the consistency model of the artifactory storage? If there is any possibility that reading the lock data here could fail without a delay, then this is absolutely not suitable for a lock implementation. The locks must be strongly consistent, and hence checking this data a second time should never be required.

File: backend/remote-state/artifactory/client.go [hunk 6]
```diff
@@ -48,16 +79,88 @@ func (c *ArtifactoryClient) Get() (*remote.Payload, error) {
...
+				}
+			}
+		}
+	} else {
+		return "", fmt.Errorf("Failed to lock: %v", err)
+	}
+}
+
+func (c *ArtifactoryClient) Unlock(id string) error {
+	if c.lockUrl == "" {
```
Comment (jbardin) [inline backend/remote-state/artifactory/client.go]: The current contract requires that an incorrect ID fail to unlock the backend. It doesn't appear to be checking the ID value here.

File: backend/remote-state/artifactory/client_test.go [hunk 7]
```diff
@@ -52,4 +60,28 @@ func TestArtifactoryFactory(t *testing.T) {
...
+	}
+	if artifactoryClient.unlockNativeClient.Config.Password != "unlock-testpass" {
+		t.Fatalf("Incorrect unlock_password was populated")
+	}
+	if artifactoryClient.lockRepo != "terraform-lock-repo" {
+		t.Fatalf("Incorrect lock_repo was populated")
+	}
+	if artifactoryClient.lockSubpath != "lock-myproject" {
+		t.Fatalf("Incorrect lock_subpath was populated")
+	}
```
Comment (jbardin) [inline backend/remote-state/artifactory/client_test.go]: Backend tests that exercise locking and concurrency also need to be added.

File: backend/remote-state/artifactory/backend.go (hunk 1 above)
Comment (bzcnsh) [reply backend/remote-state/artifactory/backend.go]: Hi jbardin,

Thanks for spending the time on this PR beyond explaining the situation. I really appreciate you trying to make this useful!

Best!

**Pull request closed without merging**